LFS-670: As a user I can upload a file linked to a subject

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-670)

Test branch: ` LFS-670_LFS-675-test`

To test: Create a new Demopgraphics form. At the bottom, a new kind of question is added where you can upload a file. Uploading a file with .txt will cause it to fill out family_history with the parsed filename.

OrdiNeu committed:
LFS-670 / LFS-675: New Answer type: FileResourceAnswer

OrdiNeu committed:
LFS-670 / LFS-676: New Answer Component: FileResourceQuestion

[wip] Move hardcoded elements in dragAndDrop.jsx to arguments, added
them to FileResourceQuestion, added FileResourceQuestion

OrdiNeu committed:
[wip] Test question

OrdiNeu committed:
[wip] testing stuff, need to switch branches

OrdiNeu committed:
LFS-670 / LFS-675: [wip] Add the file upload to a static directory

OrdiNeu committed:
LFS-670 / LFS-675: [wip] Add support for arbitrary changing of answer
fields

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Uploading to a FileResourceQuestion with a namePattern allows you to
fill out other questions from a file upload.

OrdiNeu committed:
LFS-670: [wip] Further fixes to name parsing

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject
Fix the parsing of variables extracted from filenames

OrdiNeu committed:
[wip] Remove debug code

IntegralProgrammer commented:
The "Save" button should enter the "Save" state when a file is uploaded. Instead, no state transition occurs.

- Create a new _Demographics Data_ form
- The "Save" button displays "Save"
- Set a _Family_ name
- The "Save" button still displays "Save"
- Click "Save"
- The "Save" button now displays "Saved"
- Specify a Gender
- The "Save" button now displays "Save"
- Click "Save"
- The "Save" button now displays "Saved"
- Upload a file
- The "Save" button still displays "Saved" even though one would expect it to display "Save" and there is file upload metadata that still needs to be saved.

IntegralProgrammer commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -210,6 +210,8 @@
   - chromosomeMT (boolean)
 
+  // For file answers, a string specifying if the filename can be processed to extract the subject ID from it.
+
```
Should there be a CND property specified here? There is a newly added comment but no code added here.

sdumitriu replied:
Same comment as Michael.

OrdiNeu replied:
Fixed in `f0a100a` :heavy_check_mark:

sashaandjic commented:
I uploaded 3 files: a.txt, b.txt and c.txt and saved. Family history field was correctly updated and all 3 files were listed at the bottom of the form.
Then I deleted the b.txt from Family history field and saved the form.
At the bottom of the form I can still see b.txt being listed as an uploaded file.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -239,27 +240,29 @@ function Form (props) {
-              .filter(([key, value]) => ENTRY_TYPES.includes(value['jcr:primaryType']))
-              .map(([key, entryDefinition]) => <FormEntry key={key} entryDefinition={entryDefinition} path={"."} depth={0} existingAnswers={data} keyProp={key} classes={classes} onChange={()=>setLastSaveStatus(undefined)}></FormEntry>)
-          }
+          <FormUpdateProvider>
```
What is the difference between `FormProvider` and `FormUpdateProvider`, and why do we need both at the same time?

OrdiNeu replied:
FormProvider gives the contents of the form (Answers, etc.) as a dictionary of question name -> array of answers.
FormUpdateProvider allows form components to request to update one another, via a dictionary of question name ->array of new answers to be added.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -190,6 +191,36 @@ function MultipleChoice(props) {
+    updatedOptions.forEach((option) => {
+      if (isRadio) {
+        selectOption(option, option);
+      } else if (maxAnswers !== 1) {
```
What happens if this is not a radio, and maxAnswers is 1?

OrdiNeu replied:
Resolved in `c7fd396` -- it now changes the bare input to the value of the last file :heavy_check_mark:

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,234 @@
+  enableUnknown: PropTypes.bool,
+  yesLabel: PropTypes.string,
+  noLabel: PropTypes.string,
+  unknownLabel: PropTypes.string
```
Leftover copy/paste?

OrdiNeu replied:
Fixed in `3e4ab760` :heavy_check_mark:

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,234 @@
+export default StyledFileResourceQuestion;
+
+AnswerComponentManager.registerAnswerComponent((questionDefinition) => {
+  if (questionDefinition.dataType === "file") {
```
I think a new line added to [dataentry.cnd](https://github.com/ccmbioinfo/lfs/blob/dev/modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd#L167) would be nice.

OrdiNeu replied:
There's already an entry for FileResourceAnswers in there -- did you mean something else?

sdumitriu replied:
I mean a description for the new possible value in the comments for `lfs:Question / dataType`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,234 @@
+    let data = new FormData();
+    data.append(file['name'], file);
+    data.append('jcr:primaryType', 'nt:folder');
+    return fetch("/Uploads/", {
```
This will be problematic with rights in other than the default mode (permissions-open). When the file resides under the Form node itself, it automatically inherits the rights of the form, but in a separate space we will have to explicitly synchronize the rights of the file with those of the parent form. Can you remind me why the files need to be in a different place?

OrdiNeu replied:
Fixed in 8697a73 :heavy_check_mark:

sdumitriu reviewed (changes requested):
Using the questionnaire editor, added a new question with data type "file", created a new form, clicked on the upload box, selected 2 files, pressed OK, blank screen from this error:

> Uncaught TypeError: varNames is undefined
    FileResourceQuestion webpack:///./src/questionnaire/FileResourceQuestion.jsx?:249

OrdiNeu committed:
LFS-670 / LFS-675: New Answer type: FileResourceAnswer

Fix the namePattern variable

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix a possible variable error when namePattern is unspecified

sashaandjic commented:
I can upload many files at once, however only one value from one file makes it to the field.
The problem that Mike reported is still present. If the label on the SAVE button is "SAVED", uploading the txt file will not change it to "SAVE", so to the user it appears as the change is already saved.

sashaandjic commented:
The other problem is that now with the file upload I can add multiple values to single value fields.
![image](https://user-images.githubusercontent.com/23243244/100812158-55e01c00-340a-11eb-9a07-662a76aeaeb7.png)
![image](https://user-images.githubusercontent.com/23243244/100812197-7019fa00-340a-11eb-89a9-26d2744b5ef0.png)

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix the auto-form-fillin when the input is a bare input

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix the PropTypes of FileResourceQuestion

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Make the uploaded file a child of the FileResourceQuestion, instead of
uploading to a static folder

sashaandjic commented:
- upon loading a txt file the SAVED button does not toggle to SAVE
- the previously uploaded files are no longer shown at the bottom of the form in this scenario:
  **step1**: upload file and save --> the Family history field will be updated and the file will be listed at the bottom of the form
  **step2**: close form
  **step3**: open the form again
  **step4**:  click SAVE bttn ( the button label will turn into SAVED as expected)
  **step5**: upload another file and observe that another value has been added to the Family history and another file is listed at the bottom of the form, SAVED button did not turn into SAVE (**wrong**) 
  **step6**. close the form without saving it and open it again -> this time Family field will not have the newly added value as it was not saved, however no files will be shown at the bottom of the form ( **wrong**)

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject
Force the form to re-allow saving after file uploads are complete

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix an issue when reloading the page with an uploaded file

marta- committed:
Merge branch 'dev' into LFS-670_LFS-675

sashaandjic reviewed (approved):
**Tested and approved** under the assumption that when uploading multiple files, only one will be taken into account as shown below:
![image](https://user-images.githubusercontent.com/23243244/101835643-11d7d000-3b0a-11eb-85f6-65eb72311885.png)

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Add a "delete" button to file resource questions

OrdiNeu committed:
Merge branch 'LFS-670_LFS-675' of github.com:ccmbioinfo/lfs into LFS-670_LFS-675

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Handle maxAnswers == 1, and display a warning if more files are uploaded
than maxAnswers

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix the error display and replace the close button with delete

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix styling of the answer field

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix a bug where adding more than one option in a MultipleChoice in a single upload would only add the last one.

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix an issue where the warning would show up with 0 max/min answers

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Fix the spacing in the UI

marta- reviewed (approved):
Looks great to me

sashaandjic reviewed (approved):
tested

marta- committed:
Merge branch 'dev' into LFS-670_LFS-675

marta- committed:
Merge branch 'dev' into LFS-670_LFS-675

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,320 @@
+
+    // Match each of the values into our default initialParsedAnswers
+    initialValues.forEach((filename) => {
+      let results = [...filename[0].matchAll(nameRegex)].map((match) => match[1]);
```
This seems wrong to me. Say, for example, name pattern is `@{patient}_@{tumor}\.csv` and filename is `p01_liver.csv` then:
- `varNames` is `["patient", "tumor"]` (correct)
- `nameRegex` is `(.*)_(.*)\.csv` (fragile, but correct)
- `filename.matchAll(nameRegex)` produces one match, not two, and `match[1]` will be the first captured group, `p01`, thus `results` is `["p01"]` (wrong).

```suggestion
      let results = filename[0].match(nameRegex).slice(1);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,320 @@
+    var varNames = [...namePattern.matchAll(varNamesRegex)].map((match) => match[1]);
+
+    var clearedNamesRegex = namePattern.replaceAll(varNamesRegex, "(.+)");
+    var nameRegex = new RegExp(clearedNamesRegex, "g");
```
```suggestion
    var nameRegex = new RegExp(clearedNamesRegex);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,320 @@
+    // Determine whether or not the filename matches the namePattern (if given)
+    if (namePattern) {
+      // Regex out variable names from the namePattern
+      var results = [...file['name'].matchAll(nameRegex)].map((match) => match[1]);
```
```suggestion
      var results = file['name'].match(nameRegex).slice(1);
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,320 @@
+      ) }
+      {warning}
+      <DragAndDrop
+        accept={"*.csv"}
```
I think this shouldn't be hardcoded here, either remove or use a question setting to define it.

OrdiNeu replied:
Fixed in `c7789589` :heavy_check_mark:

sdumitriu reviewed (changes requested):
Data loss:

- upload a file -> file is stored
- save -> file is still stored
- delete that file -> file is gone
- upload the same file again -> file is stored
- save -> file is gone (wrong)

This is caused by the fact that the toDelete keeps asking to delete the file. Changing to instant delete (clicking the trash icon deletes the file even if Save is not pressed) leads to changes being persisted even without pressing save, which may be bad. However, given that in the current dev save always happens when closing the form, it is not really a bad thing; it is unavoidable, and this way we're just speeding things up.

Another issue: after selecting a file to be uploaded, every time the form is saved, the file is sent to the server again, even if it is not actually stored (it is sent as a `*` parameter, which means that it just ends up discarded on the server).

sdumitriu committed:
LFS-670: As a user I can upload a file linked to a subject
Fixed namePattern parsing to support more than one parameter

sdumitriu committed:
LFS-670: As a user I can upload a file linked to a subject
Added more explanations in the CND

sdumitriu committed:
LFS-670: As a user I can upload a file linked to a subject
Bugfix: attaching a file that doesn't follow namePattern fails

sdumitriu reviewed (changes requested):
A few more errors:

- upload a file that doesn't follow namePattern -> "something is undefined" JS error caused by assuming that the filename parsing results in a valid `results` variable (I already pushed a fix for this)
- try to define a namePattern targeting a boolean (or numeric) field; upload a file that doesn't have a valid number in that place; save -> 500 server error, which triggers a login dialog, which will keep showing up every time it is submitted
- define a namePattern, upload a file that follows it, then upload a file that doesn't follow it; at this point, in the file table the first file has the correct values listed, while the second one has empty values; delete the first file -> the extracted parameters are copied into the remaining file fields (wrong), but the input hints are also displayed, overlapping the values (also wrong)

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FileResourceQuestion.jsx:
```diff
@@ -0,0 +1,322 @@
+    data.append('jcr:primaryType', 'lfs:FileResourceAnswer');
+    data.append('question', props.questionDefinition['jcr:uuid']);
+    data.append('question@TypeHint', "Reference");
+    return fetch(outURL, {
```
Needs to be replaced with `fetchWithReLogin`.

OrdiNeu replied:
Fixed in `e3953ed` :heavy_check_mark:

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Switch from collecting a list of things to delete to instead delete the
file right away. Fix a bug where you cannot upload the same file
multiple times.

OrdiNeu commented:
> Data loss:
@sdumitriu Fixed in `0435c33` :heavy_check_mark:

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Remove the .accept property, since we do not use it

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Use fetchWithRelogin instead of fetch

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Remove bugs related to uploading the same file multiple times

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Change the patient chart to link to the file instead of displaying a
path

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Make the patient chart handle multiple values better

OrdiNeu committed:
LFS-670: As a user I can upload a file linked to a subject

Force the file links to be download links

marta- reviewed (approved):
Looks good to me (again)

sashaandjic commented:
I uploaded 10 files into **File upload test (unlimited answers, affects family history)** field of the test branch. Everything looked perfect. I could see 10 uploaded fiels and 10 values entered in the field. I saved the form and closed it.
When I reopened the form, the filed that used to have 10 values entered, had only one (the last one).

marta- commented:
> I uploaded 10 files into **File upload test (unlimited answers, affects family history)** field of the test branch. Everything looked perfect. I could see 10 uploaded fiels and 10 values entered in the field. I saved the form and closed it.
> When I reopened the form, the filed that used to have 10 values entered, had only one (the last one).

This is important to know, but not a showstopper and can be fixed in the next iteration

k-pace commented:
tested. I was able to upload multiple forms of one type/source (norav, vmax, polar) and store visit date which mimics the use case for the cardiac rehab clinic

sashaandjic reviewed (approved):
tested - I haven't found any problems except the one reported above
